canonicalwebteam/docs/models.py

```diff
@@ -0,0 +1,297 @@
+import re
+from urllib.parse import urlparse
+
+from jinja2 import Template
+from requests.exceptions import HTTPError
+
+import dateutil.parser
+import humanize
+from bs4 import BeautifulSoup
+from canonicalwebteam.http import CachedSession
+
+# Constants
+DEFAULT_SESSION = CachedSession(expire_after=300, old_data_on_error=True)
+
+
+class RedirectFoundError(HTTPError):
+    """
+    If we encounter redirects from Discourse, we need to take action
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        url_parts = urlparse(self.response.headers["Location"])
+        self.redirect_path = re.sub("/t(/.*).json", r"\1", url_parts.path)
+
+
+class NotInCategoryError(HTTPError):
+    """
+    If we encounter redirects from Discourse, we need to take action
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        url_parts = urlparse(self.response.headers["Location"])
+        self.redirect_path = re.sub("/t(/.*).json", r"\1", url_parts.path)
+
+
+class NavigationParseError(Exception):
+    """
+    Indicates a failure to extract the navigation from
+    the frontpage content
+    """
+
+    def __init__(self, document, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.document = document
+
+
+# Private helper functions
+# ===
+
+
+def _process_html(html):
+    """
+    Post-process the HTML output from Discourse to
+    remove 'NOTE TO EDITORS' sections.
+
+    We expect these sections to be of the HTML format:
+
+    <aside class="quote no-group">
+      <blockquote>
+        <p>
+          <img title=":construction:" class="emoji" ...>
+          <strong>NOTE TO EDITORS</strong>
+          <img title=":construction:" class="emoji" ...>
+        </p>
+        <p> ... </p>
+      </blockquote>
+    </aside>
+    """
+
+    soup = BeautifulSoup(html, features="html.parser")
+    notes_to_editors_text = soup.find_all(text="NOTE TO EDITORS")
+
+    soup = _replace_notifications(soup)
+
+    for text in notes_to_editors_text:
+        # If this section is of the expected HTML format,
+        # we should find the <aside> container 4 levels up from
+        # the "NOTE TO EDITORS" text
+        container = text.parent.parent.parent.parent
+
+        if container.name == "aside" and "quote" in container.attrs["class"]:
+            container.decompose()
+
+    return str(soup)
+
+
+def _replace_notifications(soup):
+    """
+    Given some BeautifulSoup of a document,
+    replace blockquotes with the appropriate notification markup
+    """
+
+    with open("templates/docs/_notification.html") as notification_file:
```

Comment:
This template file is only used by this code, and is somewhat tightly coupled to this code. So just in this one case, I think that means this template file should probably be included within this module. Or we could not use templates at all and write the markup directly in the code? Up to you.

Reply:
since the html is quite small I could include directly in the code

Reply:
done